velodyne_pointcloud/include/velodyne_pointcloud/datacontainerbase.h

```diff
@@ -141,53 +146,86 @@ class DataContainerBase
     config_.fixed_frame = fixed_frame;
     config_.target_frame = target_frame;
 
-    config_.transform = fixed_frame.compare(target_frame) != 0;
-    if (config_.transform && !tf_ptr)
+    // only use somewhat resource intensive tf listener when it is necessary
+    if (!fixed_frame.empty() || !target_frame.empty())
     {
-      tf_ptr = boost::shared_ptr<tf::TransformListener>(new tf::TransformListener);
+      if (!tf_buffer)
+      {
+        tf_buffer = boost::shared_ptr<tf2_ros::Buffer>(new tf2_ros::Buffer());
+        tf_listener = boost::shared_ptr<tf2_ros::TransformListener>(new tf2_ros::TransformListener(*tf_buffer));
+      }
+    }
+    else
+    {
+      tf_listener.reset();
+      tf_buffer.reset();
     }
   }
 
   sensor_msgs::PointCloud2 cloud;
 
-  inline void vectorTfToEigen(tf::Vector3& tf_vec, Eigen::Vector3f& eigen_vec)
+  inline bool calculateTransformMatrix(Eigen::Affine3f& matrix, const std::string& target_frame,
+                                       const std::string& source_frame, const ros::Time& time)
   {
-    eigen_vec(0) = tf_vec[0];
-    eigen_vec(1) = tf_vec[1];
-    eigen_vec(2) = tf_vec[2];
-  }
-
-  inline bool computeTransformation(const ros::Time& time)
-  {
-    tf::StampedTransform transform;
+    geometry_msgs::TransformStamped msg;
     try
     {
-      tf_ptr->lookupTransform(config_.target_frame, cloud.header.frame_id, time, transform);
+      msg = tf_buffer->lookupTransform(target_frame, source_frame, time, ros::Duration(0.2));
```

Comment:
It's possible that this function could be called when no TF buffer has been created. Should be checked for existence with the shared pointer's bool semantics first.

Reply:
Normally there this shouldn't be possible because this function is just called when fixed_frame or target_frame are specified. However I can add an additional check to make the code more robust.

Reply:
Thanks. Better to write your intent in the code than leave it to chance.